broadcast/src/linked/signer/actor.rs

```diff
@@ -0,0 +1,939 @@
+            .await
+            .expect("unable to init journal");
+
+        // Replay journal
+        {
+            // Prepare the stream
+            let stream = journal
+                .replay(self.journal_replay_concurrency, None)
+                .await
+                .expect("unable to replay journal");
+            pin_mut!(stream);
+
+            // Read from the stream, which may be in arbitrary order.
+            // Remember the highest link height
+            let mut tip: Option<wire::Link> = None;
+            while let Some(msg) = stream.next().await {
+                let (_, _, _, msg) = msg.expect("unable to decode journal message");
+                let link = wire::Link::decode(msg).expect("journal message is unexpected format");
+                let height = link.chunk.as_ref().unwrap().height;
+                if tip.is_none() || height > tip.as_ref().unwrap().chunk.as_ref().unwrap().height {
+                    tip = Some(link);
+                }
+            }
+
+            // Set the tip
+            if let Some(link) = tip.take() {
+                let is_new = self.tip_man.put(&link);
+                assert!(is_new);
+            }
+        }
+
+        // Store journal
+        self.journals.insert(sequencer.clone(), journal);
+    }
+
+    /// Write a link to the appropriate journal.
+    ///
+    /// The journal must already be open and replayed.
+    async fn journal_append(&mut self, link: &wire::Link) {
+        let chunk = link.chunk.as_ref().unwrap();
+        let section = self.get_journal_section(chunk.height);
+        self.journals
+            .get_mut(&chunk.sequencer)
+            .expect("journal does not exist")
+            .append(section, link.encode_to_vec().into())
+            .await
+            .expect("unable to append to journal");
+    }
+
+    /// Syncs (ensures all data is written to disk) and prunes the journal for the given sequencer and height.
+    async fn journal_sync(&mut self, sequencer: &PublicKey, height: u64) {
+        let section = self.get_journal_section(height);
+
+        // Get journal
+        let journal = self
+            .journals
+            .get_mut(sequencer)
+            .expect("journal does not exist");
+
+        // Sync journal
+        journal.sync(section).await.expect("unable to sync journal");
+
+        // Prune journal, ignoring errors
+        let _ = journal.prune(section).await;
+    }
+
+    ////////////////////////////////////////
+    // Epoch
+    ////////////////////////////////////////
+
+    /// Updates the epoch to the value of the coordinator, and sets the refresh epoch deadline.
+    /// Prunes old evidence if moving to a new epoch.
+    fn refresh_epoch(&mut self) {
+        // Set the refresh epoch deadline
+        self.refresh_epoch_deadline = Some(self.runtime.current() + self.refresh_epoch_timeout);
+
+        // Ensure epoch is not before the current epoch
+        let epoch = self.coordinator.index();
```

Comment:
I wonder if it makes sense to have the application push epoch changes 🤔 .

I think a pull-based model should work fine but this just popped into my head.

Reply:
For visibility: we talked and decided to leave as-is for now. It can be changed to a push-based model later if needed